.github/workflows/build-and-test.yaml

```diff
@@ -43,56 +47,69 @@ jobs:
         # Reverse lookup for artifact base index to its architecture
         ./scripts/get-runners.py
     outputs:
-      bases: ${{ steps.charm-to-runner.outputs.bases }}
+      platforms: ${{ steps.charm-to-runner.outputs.platforms }}
 
   build:
-    name: 'Build ${{ matrix.arch.name }}-${{ matrix.arch.arch }}'
+    name: 'Build ${{ matrix.platform.charm_name }}-${{ matrix.platform.arch }}'
     strategy:
       fail-fast: false
       matrix:
-        arch: ${{ fromJSON(needs.collect-bases.outputs.bases) }}
+        platform: ${{ fromJSON(needs.collect-platforms.outputs.platforms) }}
     needs:
-      - collect-bases
+      - collect-platforms
       - static-and-unit-tests
     env:
-      PIP_CACHE_DIR: /tmp/cache/pip
-    runs-on: ["self-hosted", "linux", "${{ matrix.arch.arch == 'amd64' && 'X64' || 'ARM64' }}", "jammy", "large"]
+      CRAFT_SHARED_CACHE: /tmp/charmcraft
+    runs-on: ["self-hosted", "linux", "${{ matrix.platform.arch == 'amd64' && 'X64' || 'ARM64' }}", "jammy", "large"]
     timeout-minutes: 60
     steps:
       - name: Checkout
         uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4
-      - name: Install charmcraft
-        run: sudo snap install --classic --channel=latest/stable charmcraft
+      - name: Setup LXD
+        uses: canonical/setup-lxd@54a5806e490d92e207b57183cf111ed54bbdeff4 # v0.1.2
+        with:
+          channel: 5.21/stable
+      - name: Install charmcraft and charmcraftcache
+        run: |
+          sudo snap install --classic --channel=latest/stable charmcraft
+          sudo apt install -y pipx
+          pipx install charmcraftcache
```

Comment:
We must pin what we install via pip always.

Reply:
+1, make sense to me.